tokio-util/src/codec/framed.rs

```diff
@@ -17,11 +18,38 @@ use std::task::{Context, Poll};
 /// the `Encoder` and `Decoder` traits to encode and decode frames.
 ///
 /// You can create a `Framed` instance by using the `AsyncRead::framed` adapter.
+#[pin_project]
 pub struct Framed<T, U> {
+    #[pin]
     inner: FramedRead2<FramedWrite2<Fuse<T, U>>>,
 }
 
-pub(crate) struct Fuse<T, U>(pub(crate) T, pub(crate) U);
+#[pin_project]
+pub(crate) struct Fuse<T, U> {
+    #[pin]
+    pub(crate) io: T,
+    pub(crate) codec: U,
+}
+
+pub(crate) trait ProjectFuse {
```

Comment:
Why this trait is needed? 
(FYI: [the projection type generated by pin-project has basically the same visibility as the original type.](https://github.com/taiki-e/pin-project/blob/c75d010eed7bf8146dd2f5228ed66c3cd06d50e9/pin-project-internal/src/lib.rs#L63-L65))

Reply:
It abstracts over `FramedRead2` being either `FramedRead2<FramedWrite2<Fuse<T, U>>>` or `FramedRead2<Fuse<T, U>>` and lets the `io` and `codec` parts be extracted either way.

Without it `encode/decode` can't be called due to `Self` being pinned.

Reply:
Oh, that makes sense. Thanks for clarifying.

Reply:
I think it could be helpful to add a comment here explaining that?